### PR TITLE
feat: added `networkInterceptor` hook

### PR DESCRIPTION
**Pull request by IamMayankThakur**
## Summary

this hook can be used to capture all requests sent to the core

-   Added `networkInterceptor` to the `TypeInput` config.
  - This can be used to capture/modify all the HTTP requests sent to the core.

## Related issues

-   https://github.com/supertokens/supertokens-core/issues/865

## Test Plan

Added unit tests

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [x] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [x] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [x] Had run `npm run build-pretty`
-   [x] Had installed and ran the pre-commit hook
-   [ ] If new thirdparty provider is added,
    -   [ ] update switch statement in `recipe/thirdparty/providers/configUtils.ts` file, `createProvider` function.
    -   [ ] add an icon on the user management dashboard.
-   [x] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `add-ts-no-check.js` file to include that
-   [ ] If added a new recipe / api interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [ ] If added a new recipe, then make sure to expose it inside the recipe folder present in the root of this repo. We also need to expose its types.

## Remaining TODOs for this PR

-   [ ] Documentation

### CODE REVIEW

File: lib/ts/querier.ts [hunk 1]
```diff
@@ -270,6 +352,23 @@ export class Querier {
                         rid: this.rIdToCore,
                     };
                 }
+                if (Querier.networkInterceptor !== undefined) {
+                    let request = Querier.networkInterceptor(
+                        {
+                            url: url,
+                            method: "put",
+                            headers: headers,
+                            params: undefined,
```
Comment (rishabhpoddar) [inline lib/ts/querier.ts:361]:
```suggestion
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/types.ts [hunk 2]
```diff
@@ -56,6 +57,16 @@ export type TypeInput = {
     debug?: boolean;
 };
 
+export type NetworkInterceptor = (request: HttpRequest, userContext: any) => HttpRequest;
+
+export interface HttpRequest {
+    url: string;
+    method: HTTPMethod;
+    headers: { [key: string]: string | number | string[] };
+    params?: Record<string, boolean | number | string | undefined>;
```
[the remark this answers is not in the document]
Comment (IamMayankThakur) [reply lib/ts/types.ts:66]: To align the types with the existing type for `params` in the querier.

https://github.com/supertokens/supertokens-node/blob/e9ba8f8d4906387bce77d3187af694aa659bed78/lib/ts/querier.ts#L182

File: lib/ts/recipe/session/sessionFunctions.ts [hunk 3]
```diff
@@ -33,6 +33,7 @@ export async function createNewSession(
     tenantId: string,
     recipeUserId: RecipeUserId,
     disableAntiCsrf: boolean,
+    userContext: any,
```
Comment (rishabhpoddar) [inline lib/ts/recipe/session/sessionFunctions.ts:36]: should be the last param, and non optional.. please check in all places! So you can remove the `= {}` from the other two params of this function, and manually check all places where this function is being called to make sure that the params are correctly provided to the function call.

File: lib/ts/recipe/session/sessionFunctions.ts [hunk 4]
```diff
@@ -384,17 +401,22 @@ export async function revokeAllSessionsForUser(
     helpers: Helpers,
     userId: string,
     revokeSessionsForLinkedAccounts: boolean,
+    userContext: any,
```
Comment (rishabhpoddar) [inline lib/ts/recipe/session/sessionFunctions.ts:404]: should be the last param.. and non optional.

File: lib/ts/recipe/session/sessionFunctions.ts [hunk 5]
```diff
@@ -405,39 +427,56 @@ export async function getAllSessionHandlesForUser(
     helpers: Helpers,
     userId: string,
     fetchSessionsForAllLinkedAccounts: boolean,
+    userContext: any,
```
Comment (rishabhpoddar) [inline lib/ts/recipe/session/sessionFunctions.ts:430]: should be the last param.. and non optional

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/session/sessionFunctions.ts [hunk 6]
```diff
@@ -401,13 +401,10 @@ export async function revokeAllSessionsForUser(
     helpers: Helpers,
     userId: string,
     revokeSessionsForLinkedAccounts: boolean,
-    userContext: any,
-    tenantId?: string,
-    revokeAcrossAllTenants?: boolean
+    tenantId: string,
+    revokeAcrossAllTenants: boolean,
```
Comment (rishabhpoddar) [inline lib/ts/recipe/session/sessionFunctions.ts:405]:
```suggestion
    tenantId: string | undefined,
    revokeAcrossAllTenants: boolean | undefined,
```

File: lib/ts/recipe/session/sessionFunctions.ts [hunk 7]
```diff
@@ -427,13 +424,10 @@ export async function getAllSessionHandlesForUser(
     helpers: Helpers,
     userId: string,
     fetchSessionsForAllLinkedAccounts: boolean,
-    userContext: any,
-    tenantId?: string,
-    fetchAcrossAllTenants?: boolean
+    tenantId: string,
+    fetchAcrossAllTenants: boolean,
```
Comment (rishabhpoddar) [inline lib/ts/recipe/session/sessionFunctions.ts:428]:
```suggestion
    tenantId: string | undefined,
    fetchAcrossAllTenants: boolean | undefined,
```

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**